Clean unnecessary requires

zenspider opened the pull request:
Partially dealing with #4133. This removes a bunch of very heavy requires from test/helper.rb and changes the way that resources are included at the top level. Now resources and tests can require what they directly depend on and not bring in everything all in one very expensive go.

+ master:
  + 218 tests in 159s
  + require "inspec/base_cli" = true in 1.7849
+ clean_unnecessary_requires:
  + 218 tests in 98s
  + require "inspec/utils/telemetry" = true in 0.4844

zenspider committed:
Removed aws resource requiring from test/helper and inspec/resource.

This speeds up parallel unit test runs from a very consistent 2:49 to
a very consistent 1:53, or a 33% reduction.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Moved out all of the requires for inspec/resources/* from inspec/resource.rb

All requires moved to inspec/resources.rb

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Get user_test.rb working

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Get yaml_test.rb working

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Get all non-resource tests fixed again.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Got the rest of the tests to pass.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Stripping test/helper.rb of all the big inspec requires.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Got the rest of the functionals to pass.

Signed-off-by: Ryan Davis <zenspider@chef.io>

miah commented on lib/inspec/resources/filesystem.rb:
```diff
@@ -1,3 +1,5 @@
+require 'inspec/resources/command'
+
 module Inspec::Resources
   class FileSystemResource < Inspec.resource(1)
```
Can we rename this one to `FileSystem` ?

zenspider replied:
Totally... but I agree with @clintoncwolfe that we should do renames in a separate PR. I have a stash somewhere that'll report all the renames needed on first load so it should be quick and easy.

clintoncwolfe commented on lib/inspec/resources/bash.rb:
```diff
@@ -3,7 +3,7 @@
 
 module Inspec::Resources
-  class Bash < Cmd
+  class Bash < Command
```
I agree that this is nicer, but I think things like this should be done on separate PRs in the future - this is massive PR, and it's important to keep focus. This change isn't actually needed to help improve require-isolation.

zenspider replied:
I did this one because I was tired and getting confused by cmd the windows shell vs Command, esp once it started being subclassed by wholly different shells.

clintoncwolfe commented on lib/inspec/resources/command.rb:
```diff
@@ -84,4 +84,6 @@ def to_s
     end
   end
+
+  Cmd = Command
```
This sort of aliasing *definitely* makes me think a resource class rename sweep should happen on a separate PR.

clintoncwolfe commented on lib/inspec/shell.rb:
```diff
@@ -119,7 +119,7 @@ def help(topic = nil)
       elsif topic == 'matchers'
         print_matchers_help
-      elsif !Inspec::Resource.registry[topic].nil?
+      elsif !Inspec::Resource.registry[topic].nil? # TODO: fix
```
fix what? If you are leaving a note for a future developer (even you), please be kind and leave enough context to make it obvious what is problematic enough to make you leave a TODO. Thanks!

zenspider replied:
I will always know what this means. Unnecessary logic is unnecessary. I just didn't want to fix it in this PR. I have a slew of these I'd like to do soonish.

clintoncwolfe commented on test/functional/helper.rb:
```diff
@@ -8,6 +8,8 @@
 require 'yaml'
 require 'tmpdir'
 
+require 'inspec/resources' # TODO: work this back out
```
I'm actually surprised this is needed here at all - test/functional/helper is mostly worried about spawning subprocesses, not the actual internals of the InSpec resources that the subprocesses may be exercising. What needs the resources here?

zenspider replied:
`test/functional/inspec_report_test.rb` for sure. Had a few other failures but it is starting to look like they're unstable tests instead.

zenspider replied:
```
Run options: --seed 8843

# Running:

SE

Finished in 0.146366s, 13.6644 runs/s, 0.0000 assertions/s.

  1) Error:
inspec report tests::report#test_0001_loads a json report:
NoMethodError: undefined method `file' for #<#<Class:0x00007fb64ccbf5e8>:0x00007fb64ccbead0>
Did you mean?  fail
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/dsl.rb:49:in `method_missing'
    /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/old-examples/profile/controls/example.rb:21:in `load_with_context'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/profile_context.rb:162:in `instance_eval'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/profile_context.rb:162:in `load_with_context'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/profile_context.rb:146:in `load_control_file'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/profile.rb:202:in `block in collect_tests'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/profile.rb:199:in `each'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/profile.rb:199:in `collect_tests'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/runner.rb:111:in `block in load'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/runner.rb:94:in `each'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/runner.rb:94:in `load'
    /Users/ryan/Work/git/inspec/inspec/lib/inspec/runner.rb:122:in `run'
    test/functional/inspec_report_test.rb:12:in `block (3 levels) in <main>'

2 runs, 0 assertions, 0 failures, 1 errors, 1 skips

You have skipped tests. Run with --verbose for details.
        1.65 real         0.94 user         0.49 sys
```

zenspider replied:
I can push the require down since it seems to just be this one

clintoncwolfe commented on test/unit/resources/aws_billing_report_test.rb:
```diff
@@ -1,5 +1,11 @@
+require 'inspec/resource'
+require 'resources/aws/aws_billing_report'
+
+require_relative 'aws_billing_backend' # TODO: move this to mocks
```
Note - as the AWS resources are being moved to the inspec-resource pack, development has stopped on these resources - and the resource pack has a different test harness, IIRC. So, we don't need to invest too much here.

clintoncwolfe reviewed (changes requested):
I think this is s great change, and will really open up some new features like autoloading. My major concern is that some resource classes are renamed on the PR (which is an unrelated change), while others are marked as a TODO to be renamed, and another is commented.  Likewise, xinetd file was renamed to match its class. I'd prefer that this PR just be about the `require` changes, and make the renames on a separate PR.  If that seems onerous, let's at least be all-or-nothing, and do all renames.

zenspider committed:
Extra fix for utils move

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Additional whitespace cleanup.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Even more removals and moves!

Turns out we don't even use CMD in any units... so why make it?

Lots of other stuff got ripped out. Tests basically shouldn't need to
require anything but 1 implementation file and a test framework.
Anything more than that is a smell.

Signed-off-by: Ryan Davis <zenspider@chef.io>
